pkg/codegen/pcl/invoke.go

```diff
@@ -92,22 +77,140 @@ func (b *binder) bindInvokeSignature(args []model.Expression) (model.StaticFunct
+}
+
+func (b *binder) makeSignature(argsType, returnType model.Type) model.StaticFunctionSignature {
+	return model.StaticFunctionSignature{
+		Parameters: []model.Parameter{
+			{
+				Name: "token",
+				Type: model.StringType,
+			},
+			{
+				Name: "args",
+				Type: argsType,
+			},
+			{
+				Name: "provider",
+				Type: model.NewOptionalType(model.StringType),
+			},
+		},
+		ReturnType: returnType,
 	}
+}
+
+func (b *binder) zeroSignature() model.StaticFunctionSignature {
+	return b.makeSignature(model.NewOptionalType(model.DynamicType), model.DynamicType)
+}
 
-	// Create args and result types for the schema.
+func (b *binder) signatureForArgs(fn *schema.Function, args model.Expression) (model.StaticFunctionSignature, error) {
+	useOutputVersion := false
+	if fn.NeedsOutputVersion() {
+		outputVersionType := b.schemaTypeToType(fn.Inputs.InputShape)
+		regularVersionType := b.schemaTypeToType(fn.Inputs)
+		callsiteType := args.Type()
+		if regularVersionType.ConversionFrom(callsiteType) == model.NoConversion &&
+			outputVersionType.ConversionFrom(callsiteType) == model.SafeConversion {
+			useOutputVersion = true
+		}
+	}
+	if useOutputVersion {
+		return b.outputVersionSignature(fn)
+	}
+	return b.regularSignature(fn), nil
+}
+
+func (b *binder) regularSignature(fn *schema.Function) model.StaticFunctionSignature {
+	var argsType model.Type
 	if fn.Inputs == nil {
-		signature.Parameters[1].Type = model.NewOptionalType(model.NewObjectType(map[string]model.Type{}))
+		argsType = model.NewOptionalType(model.NewObjectType(map[string]model.Type{}))
 	} else {
-		signature.Parameters[1].Type = b.schemaTypeToType(fn.Inputs)
+		argsType = b.schemaTypeToType(fn.Inputs)
 	}
 
+	var returnType model.Type
 	if fn.Outputs == nil {
-		signature.ReturnType = model.NewObjectType(map[string]model.Type{})
+		returnType = model.NewObjectType(map[string]model.Type{})
 	} else {
-		signature.ReturnType = b.schemaTypeToType(fn.Outputs)
+		returnType = b.schemaTypeToType(fn.Outputs)
+	}
+
+	return b.makeSignature(argsType, model.NewPromiseType(returnType))
+}
+
+func (b *binder) outputVersionSignature(fn *schema.Function) (model.StaticFunctionSignature, error) {
+	if !fn.NeedsOutputVersion() {
+		return model.StaticFunctionSignature{}, fmt.Errorf("Function %s does not have an Output version", fn.Token)
+	}
+
+	// Given `fn.NeedsOutputVersion()==true`, can assume `fn.Inputs != nil`, `fn.Outputs != nil`.
+	argsType := b.schemaTypeToType(fn.Inputs.InputShape)
+	returnType := b.schemaTypeToType(fn.Outputs)
+	return b.makeSignature(argsType, model.NewOutputType(returnType)), nil
+}
+
+// Detects invoke calls that use an output version of a function.
+func IsOutputVersionInvokeCall(call *model.FunctionCallExpression) bool {
+	if call.Name == Invoke {
+		// Currently binder.bindInvokeSignature will assign
+		// either DynamicType, a Promise<T>, or an Output<T>
+		// for the return type of an invoke. Output<T> implies
+		// that an output version has been picked.
```

Comment:
Do you think it would it be worth using a different name for the output-form invoke intrinsic, or does that seem overcomplicated?

Reply:
I don't have the compiler pipeline in my head so difficult to judge, I could try to explore this? Where would such a change go?

The current change fits rather well in `bindInvokeSignature`, so I looked up call sites of that, but what I find is:

```	// Define the invoke function.
	b.root.DefineFunction(Invoke, model.NewFunction(model.GenericFunctionSignature(b.bindInvokeSignature)))
```

It is defining an (overloaded) function Invoke. That's not the right place to define it as `InvokeOutput` or some such it seems, since "overloading" happens elsewhere.

Do you believe things might fit better if we had a new `InvokeOutput` intrinsic?

Reply:
> Do you believe things might fit better if we had a new InvokeOutput intrinsic?

Nope, I think what you've laid out in your comment has convinced me that it wouldn't. Thanks!